db/compaction_iterator.cc

```diff
@@ -628,18 +634,23 @@ void CompactionIterator::PrepareOutput() {
 inline SequenceNumber CompactionIterator::findEarliestVisibleSnapshot(
     SequenceNumber in, SequenceNumber* prev_snapshot) {
   assert(snapshots_->size());
-  SequenceNumber prev = kMaxSequenceNumber;
-  for (const auto cur : *snapshots_) {
-    assert(prev == kMaxSequenceNumber || prev <= cur);
-    if (cur >= in && (snapshot_checker_ == nullptr ||
-                      snapshot_checker_->IsInSnapshot(in, cur))) {
-      *prev_snapshot = prev == kMaxSequenceNumber ? 0 : prev;
```

Comment:
I'm a little bit worried about potential regression bought by the new approach. In the old way, if there is no or only a few long running transactions, the for-loop will exit quickly. Now we are force every key to go through a binary search.

Reply:
it is possible with short-lived snapshots to regress:

```
comp(sec)  snapshots  with_pr
41.20449           0        0
41.64049           5        0
41.61606           0        1
43.80471           5        1
```

Reply:
Here's an update with longer-running stable benchmarks. Also I added more short-lived snapshots (1K) to try to make any problem more obvious.

```
snapshot   with_pr    flush(sec) comp(sec)
0          0          8.253438   89.926194
1000       0          10.212946  113.775399
1000       1          9.233056   117.127679
```

- CPU overhead for `NextFromInput` with no snapshots:

```
+    2.34%  rocksdb:low0     db_stress            [.] rocksdb::CompactionIterator::NextFromInput
```

- CPU overhead for `NextFromInput` with 1K short-lived snapshots before this PR:

```
+    2.67%  rocksdb:low0     db_stress            [.] rocksdb::CompactionIterator::NextFromInput
```

- CPU overhead for `NextFromInput` with 1K short-lived snapshots after this PR:

```
+    3.49%  rocksdb:low0     db_stress            [.] rocksdb::CompactionIterator::NextFromInput
```